lib/rig_inbound_gateway_web/v1/metadata_controller.ex

```diff
@@ -0,0 +1,195 @@
+      ) do
+
+    {err, data} = conn
+    |> accept_only_req_for(["application/json"])
+    |> extract_metadata()
+
+    if err === :error do
+      fail!(conn, data)
+    else
+      persist_metadata(data, connection_id)
+      send_resp(conn, :no_content, "")
+    end
+  end
+
+  # ---
+
+  defp extract_metadata(conn) do
+    {json_metadata_err, json_metadata} = extract_metadata_from_json(conn)
+    jwt_metadata = extract_metadata_from_jwt(conn)
+
+    if json_metadata_err === :error do
+      {:error, json_metadata}
+    else
+      # Merge values from metadata with the values from JWT, prioritizing JWT values
+      metadata = Map.merge(json_metadata, jwt_metadata)
+
+      # Check if metadata contains all the indexed fields; if not: send back a bad request
+      # This needs to be done here because now the values from the JWT got inserted
+      if has_required_indices?(metadata) do
+        {:ok, metadata}
+      else
+        message = """
+        Metadata doesn't contain indexed fields.
+        """
+        {:error, message}
+      end
+    end
+  end
+
+  # ---
+
+  defp extract_metadata_from_jwt(conn) do
+    with {:ok, claims} <- Map.get(conn.assigns, :auth_tokens, [])
+    |> Enum.map(fn
+      {"bearer", token} -> JWT.parse_token(token)
+      _ -> {:ok, %{}}
+    end)
+    |> Result.list_to_result()
+    |> Result.map_err(fn decode_errors -> {:error, decode_errors} end) do
+      conf = config()
+      metadata_from_jwt = conf.jwt_fields
+
+      # Merge the list of claims into a single map
+      claims = Enum.reduce(claims, fn x, y ->
+        Map.merge(x, y, fn _k, v1, v2 -> v2 ++ v1 end)
+      end)
+
+      # Get values from JWT
+      for {key, jwt_key} <- metadata_from_jwt,
+        jwt_val = claims[jwt_key],
+        into: %{},
+        do: {key, jwt_val}
+    else
+      {:error, _} ->
+        %{}
+    end
+  end
+
+  # ---
+
+  defp extract_metadata_from_json(conn) do
+    with {"application", "json"} <- content_type(conn),
+         {:ok, body, _conn} <- BodyReader.read_full_body(conn),
+         {:ok, json} <- Jason.decode(body),
+         {:parse, %{"metadata" => metadata}} <- {:parse, json} do
+      {:ok, metadata}
+    else
+      {:parse, json} ->
+        message = """
+        Expected field "metadata" is not present.
+
+        Decoded request body:
+        #{inspect(json)}
+        """
+
+        {:error, message}
+
+      error ->
+        message = """
+        Expected JSON encoded body.
+
+        Technical info:
+        #{inspect(error)}
+        """
+
+        {:error, message}
+    end
+  end
+
+  # ---
+
+  defp persist_metadata(metadata, connection_id) do
+    Logger.debug(fn -> "Metadata: " <> inspect(metadata) end)
+
+    conf = config()
+    indexed_fields = for x <- conf.indexed_metadata, do: {x, metadata[x]}
+
+    Logger.debug(fn -> "Indexed fields: " <> inspect(indexed_fields) end)
+
+    # TODO: Actually set metadata
+  end
+
+  # ---
+
+  defp has_required_indices?(metadata) do
```

Comment:
what about `has_required_fields?` instead? Without a descriptions, it's quite hard to figure out what it means for a map to "have indices".